### PR TITLE
[PDI-17423] - PDI REST Client Not Sending Content-Length(setting chun…

### PR DESCRIPTION
**Pull request by bcostahitachivantara**
…ked)

@pentaho/tatooine 
@graimundo 
@pentaho-lmartins

### CODE REVIEW

File: engine/src/main/java/org/pentaho/di/trans/steps/rest/StringMessageBodyWriter.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+/**
+ * This class is a custom provider to replace the default com.sun.jersey.core.impl.provider.entity.StringProvider
+ * which it has a bad implementation related with get size.
+ *
+ *
+ */
+@Produces( { "text/plain", "*/*" } )
+public final class StringMessageBodyWriter implements MessageBodyWriter<String> {
+
+  private StringProvider stringProvider;
```
Comment (graimundo) [inline engine/src/main/java/org/pentaho/di/trans/steps/rest/StringMessageBodyWriter.java]: can be final as well

File: engine/src/main/java/org/pentaho/di/trans/steps/rest/StringMessageBodyWriter.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import javax.ws.rs.ext.MessageBodyWriter;
+
+/**
+ * This class is a custom provider to replace the default com.sun.jersey.core.impl.provider.entity.StringProvider
+ * which it has a bad implementation related with get size.
+ *
+ *
+ */
+@Produces( { "text/plain", "*/*" } )
+public final class StringMessageBodyWriter implements MessageBodyWriter<String> {
```
Comment (graimundo) [inline engine/src/main/java/org/pentaho/di/trans/steps/rest/StringMessageBodyWriter.java]: Can the class be package private?

**Comment by buildguy** [comment]
### :warning: Build finished in 31m 28s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl engine
```
---
### :no_entry: Failed Tests

#### :cloud_with_lightning_and_rain: 1 test(s) failed:

<details><summary><b>org.pentaho.di.trans.steps.rest.RestTest.testCallEndpointWithDeleteVerb</b> (click to expand)</summary><p>

```

Can not result from [null]
 at org.apache.maven.surefire.booter.ForkedBooter.main (ForkedBooter.java:413)
 at org.apache.maven.surefire.booter.ForkedBooter.execute (ForkedBooter.java:125)
 at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess (ForkedBooter.java:340)
 at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader (ForkedBooter.java:379)
 at org.apache.maven.surefire.junit4.JUnit4Provider.invoke (JUnit4Provider.java:159)
 at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet (JUnit4Provider.java:238)
 at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun (JUnit4Provider.java:273)
 at org.apache.maven.surefire.junit4.JUnit4Provider.execute (JUnit4Provider.java:365)
 at org.powermock.modules.junit4.PowerMockRunner.run (PowerMockRunner.java:59)
 at org.powermock.modules.junit4.common.internal.impl.AbstractCommonPowerMockRunner.run (AbstractCommonPowerMockRunner.java:57)
 at org.powermock.modules.junit4.common.internal.impl.JUnit4TestSuiteChunkerImpl.run (JUnit4TestSuiteChunkerImpl.java:121)
 at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl.run (PowerMockJUnit44RunnerDelegateImpl.java:136)
 at org.junit.internal.runners.ClassRoadie.runProtected (ClassRoadie.java:45)
 at org.junit.internal.runners.ClassRoadie.runUnprotected (ClassRoadie.java:33)
 at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl$1.run (PowerMockJUnit44RunnerDelegateImpl.java:134)
 at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl.runMethods (PowerMockJUnit44RunnerDelegateImpl.java:160)
 at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl.invokeTestMethod (PowerMockJUnit44RunnerDelegateImpl.java:218)
 at org.junit.internal.runners.MethodRoadie.run (MethodRoadie.java:49)
 at org.junit.internal.runners.MethodRoadie.runTest (MethodRoadie.java:86)
 at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl$PowerMockJUnit44MethodRunner.runBeforesThenTestThenAfters (PowerMockJUnit44RunnerDelegateImpl.java:298)
 at org.powermock.modules.junit4.internal.impl.PowerMockJUnit47RunnerDelegateImpl$PowerMockJUnit47MethodRunner.executeTest (PowerMockJUnit47RunnerDelegateImpl.java:82)
 at org.powermock.modules.junit4.internal.impl.PowerMockJUnit47RunnerDelegateImpl$PowerMockJUnit47MethodRunner.evaluateStatement (PowerMockJUnit47RunnerDelegateImpl.java:107)
 at org.powermock.modules.junit4.internal.impl.PowerMockJUnit47RunnerDelegateImpl$PowerMockJUnit47MethodRunner$TestExecutorStatement.evaluate (PowerMockJUnit47RunnerDelegateImpl.java:147)
 at org.powermock.modules.junit4.internal.impl.PowerMockJUnit47RunnerDelegateImpl$PowerMockJUnit47MethodRunner.access$100 (PowerMockJUnit47RunnerDelegateImpl.java:59)
 at org.powermock.modules.junit4.internal.impl.PowerMockJUnit47RunnerDelegateImpl$PowerMockJUnit47MethodRunner.executeTestInSuper (PowerMockJUnit47RunnerDelegateImpl.java:131)
 at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl$PowerMockJUnit44MethodRunner.executeTest (PowerMockJUnit44RunnerDelegateImpl.java:310)
 at org.junit.internal.runners.MethodRoadie.runBeforesThenTestThenAfters (MethodRoadie.java:96)
 at org.junit.internal.runners.MethodRoadie$2.run (MethodRoadie.java:88)
 at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl$PowerMockJUnit44MethodRunner.runTestMethod (PowerMockJUnit44RunnerDelegateImpl.java:326)
 at org.junit.internal.runners.TestMethod.invoke (TestMethod.java:68)
 at java.lang.reflect.Method.invoke (Method.java:498)
 at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
 at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
 at sun.reflect.NativeMethodAccessorImpl.invoke0 (NativeMethodAccessorImpl.java:-2)
 at org.pentaho.di.trans.steps.rest.RestTest.testCallEndpointWithDeleteVerb (RestTest.java:118)
 at org.pentaho.di.trans.steps.rest.Rest$$EnhancerByMockitoWithCGLIB$$8367b66.callRest (<generated>:-1)
 at org.powermock.api.mockito.repackaged.MethodInterceptorFilter.intercept (MethodInterceptorFilter.java:60)
 at org.mockito.internal.handler.InvocationNotifierHandler.handle (InvocationNotifierHandler.java:38)
 at org.mockito.internal.handler.NullResultGuardian.handle (NullResultGuardian.java:29)
 at org.mockito.internal.handler.MockHandlerImpl.handle (MockHandlerImpl.java:91)
 at org.mockito.internal.stubbing.StubbedInvocationMatcher.answer (StubbedInvocationMatcher.java:34)
 at org.mockito.internal.stubbing.answers.CallsRealMethods.answer (CallsRealMethods.java:41)
 at org.mockito.internal.invocation.InvocationImpl.callRealMethod (InvocationImpl.java:112)
 at org.mockito.internal.invocation.realmethod.CleanTraceRealMethod.invoke (CleanTraceRealMethod.java:30)
 at org.mockito.internal.invocation.realmethod.DefaultRealMethod.invoke (DefaultRealMethod.java:21)
 at org.powermock.api.mockito.repackaged.DelegatingMockitoMethodProxy.invokeSuper (DelegatingMockitoMethodProxy.java:20)
 at org.mockito.cglib.proxy.MethodProxy.invokeSuper (MethodProxy.java:216)
 at org.pentaho.di.trans.steps.rest.Rest$$EnhancerByMockitoWithCGLIB$$8367b66$$FastClassByMockitoWithCGLIB$$1c1a34e7.invoke (<generated>:-1)
 at org.pentaho.di.trans.steps.rest.Rest$$EnhancerByMockitoWithCGLIB$$8367b66.CGLIB$callRest$4 (<generated>:-1)
 at org.pentaho.di.trans.steps.rest.Rest.callRest (Rest.java:119)

```

</p></details>

  
Tests run: **3631**, Failures: **1**, Skipped: **7** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/pdi/job/PR-7216/1/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

File: engine/src/main/java/org/pentaho/di/trans/steps/rest/StringMessageBodyWriter.java (hunk 1 above)
Comment (bcostahitachivantara) [reply engine/src/main/java/org/pentaho/di/trans/steps/rest/StringMessageBodyWriter.java]: It cannot be final to this class be able to be mocked.

**Comment by bcostahitachivantara** [comment]
> Missing unit tests.
> Can you check if, instead of setting the writer just for this case, we can set it to be used by default in the system?

In this case we create a new client based on an configuration object where we have already added a new custom provider. I did not find any configuration file, so i think that cannot be global.

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 29m 18s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl engine
```
---
:ok_hand: All tests passed!
 
Tests run: **3632**, Failures: **0**, Skipped: **7** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/pdi/job/PR-7216/2/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 30m 13s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl engine
```
---
:ok_hand: All tests passed!
 
Tests run: **3634**, Failures: **0**, Skipped: **7** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/pdi/job/PR-7216/3/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

File: engine/src/main/java/org/pentaho/di/trans/steps/rest/StringMessageBodyWriter.java (hunk 1 above)
Comment (graimundo) [reply engine/src/main/java/org/pentaho/di/trans/steps/rest/StringMessageBodyWriter.java]: Now that the UTs were fixed, this can be final.

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 30m 46s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl engine
```
---
:ok_hand: All tests passed!
 
Tests run: **3634**, Failures: **0**, Skipped: **7** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/pdi/job/PR-7216/4/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Pull request merged**